provider/reprovider.go

```diff
@@ -0,0 +1,137 @@
+package provider
+
+import (
+	"context"
+	"gx/ipfs/QmS2aqUZLJp8kF1ihE5rvDGE5LvmKDPnx32w9Z1BW9xLV5/go-ipfs-blockstore"
+	"gx/ipfs/QmZBH87CAPFHcc7cYmBqeSQ98zQ3SX9KUxiYgzPmLWNVKz/go-libp2p-routing"
+	"time"
+)
+
+var (
+	reprovideOutgoingWorkerLimit = 8
+)
+
+type doneFunc func(error)
+
+type Reprovider struct {
+	ctx context.Context
+	queue *Queue
+	tracker *Tracker
+	tick time.Duration
+	blockstore blockstore.Blockstore
+	contentRouting routing.ContentRouting
+	trigger chan doneFunc
+}
+
+// Reprovider periodically re-announces the cids that have been provided. These
+// reprovides can be run on an interval and/or manually. Reprovider also untracks
+// cids that are no longer in the blockstore.
+func NewReprovider(ctx context.Context, queue *Queue, tracker *Tracker, tick time.Duration, blockstore blockstore.Blockstore, contentRouting routing.ContentRouting) *Reprovider {
+	return &Reprovider{
+		ctx: ctx,
+		queue: queue,
+		tracker: tracker,
+		tick: tick,
+		blockstore: blockstore,
+		contentRouting: contentRouting,
+		trigger: make(chan doneFunc),
+	}
+}
+
+// Begin listening for triggers and reprovide whatever is
+// in the reprovider queue.
+func (rp *Reprovider) Run() {
+	go rp.handleTriggers()
+	go rp.handleAnnouncements()
+}
+
+// Add all the cids in the tracker to the reprovide queue
+func (rp *Reprovider) Reprovide() error {
+	cids, err := rp.tracker.Tracking(rp.ctx)
+	if err != nil {
+		return err
+	}
+	for c := range cids {
+		if err := rp.queue.Enqueue(c); err != nil {
+			log.Warningf("unable to enqueue cid: %s, %s", c, err)
+			continue
+		}
+	}
+	return nil
+}
+
+// Trigger causes a reprovide
+func (rp *Reprovider) Trigger(ctx context.Context) error {
+	progressCtx, cancel := context.WithCancel(ctx)
+
+	var err error
+	done := func(e error) {
+		err = e
+		cancel()
+	}
+
+	select {
+	case <-rp.ctx.Done():
+		return context.Canceled
+	case <-ctx.Done():
+		return context.Canceled
+	case rp.trigger <- done:
+		select {
+		case <-progressCtx.Done():
+			return err
+		case <-ctx.Done():
+			return context.Canceled
+		}
+	}
+}
+
+func (rp *Reprovider) handleTriggers() {
+	// dont reprovide immediately.
+	// may have just started the daemon and shutting it down immediately.
+	// probability( up another minute | uptime ) increases with uptime.
+	after := time.After(time.Minute)
+	var done doneFunc
+	for {
+		if rp.tick == 0 {
+			after = make(chan time.Time)
```

Comment:
This is just a channel that always blocks right? You can just use `nil` if you want.

Reply:
Probably! I moved this code over, let me try that out.

Reply:
Updated.